### PR TITLE
[#1468] feat(jdbc): Obtain MySQL table information in the form of metadata using the driver.

### PR DESCRIPTION
**Pull request by Clearvive**
### What changes were proposed in this pull request?
Obtain MySQL table information in the form of metadata using the driver.

### Why are the changes needed?

Fix: #1468

### Does this PR introduce _any_ user-facing change?
NO

### How was this patch tested?
UT/IT

### CODE REVIEW

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java [hunk 1]
```diff
@@ -572,10 +521,11 @@ private StringBuilder appendColumnDefinition(JdbcColumn column, StringBuilder sq
       sqlBuilder.append("NOT NULL ");
     }
     // Add DEFAULT value if specified
-    // TODO: uncomment this once we support column default values.
-    // if (StringUtils.isNotEmpty(column.getDefaultValue())) {
-    //   sqlBuilder.append("DEFAULT '").append(column.getDefaultValue()).append("'").append(SPACE);
-    // }
+    // TODO #1556 will add default value.
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:524]: suggest to remove
```
    //    if (StringUtils.isNotEmpty(column.getDefaultValue())) {
    //      sqlBuilder.append("DEFAULT
    // '").append(column.getDefaultValue()).append("'").append(SPACE);
    //    }
```

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java [hunk 2]
```diff
@@ -48,46 +43,95 @@ public class MysqlTableOperations extends JdbcTableOperations {
...
-              .withNullable(nullable)
-              .withComment(comment)
-              // TODO: uncomment this once we support column default values.
-              // .withDefaultValue("NULL".equals(defaultValue) ? null : defaultValue)
-              .withProperties(properties)
-              .build());
+    try (Connection connection = getConnection(databaseName)) {
+      // 1.Get table information
+      DatabaseMetaData metaData = connection.getMetaData();
+      ResultSet table = metaData.getTables(databaseName, null, tableName, null);
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:50]: why `schemaPattern` is null when `getTables`? I wonder maybe getting multi-tables if not specify schemaPattern;

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java [hunk 3]
```diff
@@ -48,46 +43,95 @@ public class MysqlTableOperations extends JdbcTableOperations {
...
+      if (!table.next() || !tableName.equals(table.getString("TABLE_NAME"))) {
+        throw new NoSuchTableException(
+            String.format("Table %s does not exist in %s.", tableName, databaseName));
+      }
+      String comment = table.getString("REMARKS");
+
+      // 2.Get index information
+      Map<String, Set<String>> indexGroupByName = new HashMap<>();
+      ResultSet indexInfo = metaData.getIndexInfo(databaseName, null, tableName, false, false);
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:59]: miss schema in `metaData.getIndexInfo`

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java [hunk 4]
```diff
@@ -48,46 +43,95 @@ public class MysqlTableOperations extends JdbcTableOperations {
...
+      DatabaseMetaData metaData = connection.getMetaData();
+      ResultSet table = metaData.getTables(databaseName, null, tableName, null);
+      if (!table.next() || !tableName.equals(table.getString("TABLE_NAME"))) {
+        throw new NoSuchTableException(
+            String.format("Table %s does not exist in %s.", tableName, databaseName));
+      }
+      String comment = table.getString("REMARKS");
+
+      // 2.Get index information
+      Map<String, Set<String>> indexGroupByName = new HashMap<>();
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:58]: Since Index is not merged, suggest to remove get  index code

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java [hunk 5]
```diff
@@ -48,46 +43,95 @@ public class MysqlTableOperations extends JdbcTableOperations {
...
+      ResultSet indexInfo = metaData.getIndexInfo(databaseName, null, tableName, false, false);
+      while (indexInfo.next()) {
+        if (!indexInfo.getBoolean("NON_UNIQUE")) {
+          indexGroupByName.compute(
+              indexInfo.getString("COLUMN_NAME"),
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:63]: is it possbile multi  column name?

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java [hunk 6]
```diff
@@ -48,46 +43,95 @@ public class MysqlTableOperations extends JdbcTableOperations {
...
+        typeBean.setDecimalDigits(columns.getString("DECIMAL_DIGITS"));
+
+        String columnName = columns.getString("COLUMN_NAME");
+        String colComment = columns.getString("REMARKS");
+        jdbcColumns.add(
+            new JdbcColumn.Builder()
+                .withName(columnName)
+                .withType(typeConverter.toGravitinoType(typeBean))
+                .withNullable(columns.getBoolean("NULLABLE"))
+                .withComment(StringUtils.isEmpty(colComment) ? null : colComment)
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:106]: comment is null and empty are not same, we should replace `StringUtils.isEmpty(colComment) ? null : colComment` with colComment

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java [hunk 7]
```diff
@@ -48,46 +43,95 @@ public class MysqlTableOperations extends JdbcTableOperations {
...
+                // TODO #1556 will add default value.
+                //                .withDefaultValue(columns.getString("COLUMN_DEF"))
+                .withProperties(
+                    Collections.unmodifiableList(
+                        new ArrayList<String>() {
+                          {
+                            if ("YES".equalsIgnoreCase(columns.getString("IS_AUTOINCREMENT"))) {
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:113]: AUTO_INCREMENT should be processed in another PR?

**Comment by jerryshao** [comment]
@FANNG1 @mchades please help to review on this. Thanks.

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java [hunk 8]
```diff
@@ -48,46 +43,95 @@ public class MysqlTableOperations extends JdbcTableOperations {
 
   @Override
   public JdbcTable load(String databaseName, String tableName) throws NoSuchTableException {
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:45]: seems PostgreSQL could resuse most code, is it possible to put the code in JdbcTableOperations? then PG, Snowflake could reuse it.

**Comment by FANNG1** [comment]
I remember there maybe some issues If getting infomation from JDBC like:
1. update columnA comment to xx, 
2. change columnA type from float to double
3. columnA comment is dropped
this is also fixed in this PR, yes?

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java (hunk 2 above)
Comment (FANNG1) [reply catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:50]: ok

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java [hunk 9]
```diff
@@ -43,51 +35,53 @@
...
+      // 1.Get table information
+      DatabaseMetaData metaData = connection.getMetaData();
+      ResultSet table = metaData.getTables(databaseName, null, tableName, null);
+      if (!table.next() || !tableName.equals(table.getString("TABLE_NAME"))) {
+        throw new NoSuchTableException(
+            String.format("Table %s does not exist in %s.", tableName, databaseName));
+      }
+      String comment = table.getString("REMARKS");
+
+      // 4.Get column information
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:53]: not 4

**Comment by FANNG1** [comment]
LGTM

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java [hunk 10]
```diff
@@ -43,51 +35,53 @@
...
+                // TODO #1556 will add default value.
+                //                .withDefaultValue(columns.getString("COLUMN_DEF"))
+                .withProperties(Collections.emptyList())
+                .build());
+      }
+
+      return new JdbcTable.Builder()
+          .withName(tableName)
+          .withColumns(jdbcColumns.toArray(new JdbcColumn[0]))
+          .withComment(comment)
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:78]: should we remove gravitino id from comment?

**Comment by jerryshao** [comment]
@mchades please also take time to review.

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/converter/MysqlTypeConverter.java [hunk 11]
```diff
@@ -84,4 +82,38 @@ public String fromGravitinoType(Type type) {
     }
     throw new IllegalArgumentException("Not a supported type: " + type.toString());
   }
+
+  public static class MysqlTypeBean {
+    private String typeName;
+    private String columnSize;
+    private String decimalDigits;
```
Comment (mchades) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/converter/MysqlTypeConverter.java:89]: What do these two fields mean? Please add comments.

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/converter/MysqlTypeConverter.java [hunk 12]
```diff
@@ -84,4 +82,38 @@ public String fromGravitinoType(Type type) {
     }
     throw new IllegalArgumentException("Not a supported type: " + type.toString());
   }
+
+  public static class MysqlTypeBean {
```
Comment (mchades) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/converter/MysqlTypeConverter.java:86]: The `equals` method should be overrided

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/converter/MysqlTypeConverter.java [hunk 13]
```diff
@@ -7,16 +7,14 @@
 import com.datastrato.gravitino.catalog.jdbc.converter.JdbcTypeConverter;
 import com.datastrato.gravitino.rel.types.Type;
 import com.datastrato.gravitino.rel.types.Types;
-import java.util.List;
-import net.sf.jsqlparser.statement.create.table.ColDataType;
 
 /** Type converter for MySQL. */
-public class MysqlTypeConverter extends JdbcTypeConverter<ColDataType, String> {
+public class MysqlTypeConverter
+    extends JdbcTypeConverter<MysqlTypeConverter.MysqlTypeBean, String> {
```
Comment (mchades) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/converter/MysqlTypeConverter.java:13]: Why is MysqlTypeBean necessary? Can we convert to Gravitino type from string or column information directly?

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java [hunk 14]
```diff
@@ -43,51 +35,53 @@
...
+
+        String columnName = columns.getString("COLUMN_NAME");
+        String colComment = columns.getString("REMARKS");
+        jdbcColumns.add(
+            new JdbcColumn.Builder()
+                .withName(columnName)
+                .withType(typeConverter.toGravitinoType(typeBean))
+                .withNullable(columns.getBoolean("NULLABLE"))
+                .withComment(StringUtils.isEmpty(colComment) ? null : colComment)
+                // TODO #1556 will add default value.
```
Comment (mchades) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:69]: #1556 is about Iceberg default value

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java [hunk 15]
```diff
@@ -43,51 +35,53 @@
...
+        typeBean.setDecimalDigits(columns.getString("DECIMAL_DIGITS"));
+
+        String columnName = columns.getString("COLUMN_NAME");
+        String colComment = columns.getString("REMARKS");
+        jdbcColumns.add(
+            new JdbcColumn.Builder()
+                .withName(columnName)
+                .withType(typeConverter.toGravitinoType(typeBean))
+                .withNullable(columns.getBoolean("NULLABLE"))
+                .withComment(StringUtils.isEmpty(colComment) ? null : colComment)
```
Comment (mchades) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:68]: Why need case empty string to null?

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java [hunk 16]
```diff
@@ -43,51 +35,53 @@
...
+                // TODO #1556 will add default value.
+                //                .withDefaultValue(columns.getString("COLUMN_DEF"))
+                .withProperties(Collections.emptyList())
```
Comment (mchades) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:71]: The column property is always empty?

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java [hunk 17]
```diff
@@ -43,51 +35,53 @@
...
+                .build());
+      }
+
+      return new JdbcTable.Builder()
+          .withName(tableName)
+          .withColumns(jdbcColumns.toArray(new JdbcColumn[0]))
+          .withComment(comment)
+          .withProperties(Collections.emptyMap())
```
Comment (mchades) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:79]: The table property is always empty?

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java [hunk 18]
```diff
@@ -43,51 +35,53 @@
...
+          .withAuditInfo(AuditInfo.EMPTY)
```
Comment (mchades) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:80]: Always empty?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/mysql/TestMysqlTableOperations.java [hunk 19]
```diff
@@ -49,26 +46,9 @@ public void testOperationTable() {
...
-            .withType(INT)
-            .withProperties(
-                new ArrayList<String>() {
-                  {
-                    add("UNIQUE KEY");
-                  }
-                })
-            .withNullable(true)
-            .build());
+        new JdbcColumn.Builder().withName("col_3").withType(INT).withNullable(true).build());
```
Comment (mchades) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/mysql/TestMysqlTableOperations.java:51]: Should we use `ColumnImpl` here?

**Review by mchades — COMMENTED** [review]
LGTM

**Review by FANNG1 — APPROVED** [review]
